rules/prefer-default-parameters.js

```diff
@@ -115,6 +121,14 @@ const create = context => {
 			functionStack.push(node);
 		},
 		':function:exit': () => {
+			for (const [id, report] of reportedIdentifiers) {
```

Comment:
@medusalix 

We use `variable` to analysis the identifiers https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/rules/prefer-set-has.js#L143.

Reply:
Ah, even better. No need to reinvent the wheel I guess.

Reply:
I'm not sure if we can check this,

```js
			function abc(foo) {
				baz();
				foo = foo || 'bar';
				function baz() {
					console.log(foo);
				}
			}
```

In this case, `foo` is used before the assignment, but how to know?

Reply:
Yeah, I don't think we can check whether certain functions have side-effects. But we use the suggestions API, so users still have the option to ignore the fixes.

Reply:
I don't know, though users can ignore the fix, but it will still reported.

This seems the tricky part.

Reply:
`eslint-utils` has a function called [`hasSideEffect`](https://eslint-utils.mysticatea.dev/api/ast-utils.html#hassideeffect), maybe we can check all nodes before the assignment, ignore those cases hasSideEffect?

Reply:
That works, but it also ignores this (and 5 other test cases):

```js
function abc(bar) {
	foo();
	bar = bar || 123;
}
```

It's only a heuristic, so we get lots of valid test cases that would otherwise be considered invalid.

Reply:
@sindresorhus What do you think?

Reply:
I'm fine with ignoring such cases. That code should use default parameters anyway.